pageserver/src/restore_local_repo.rs

```diff
@@ -332,12 +336,19 @@ pub fn import_timeline_wal(walpath: &Path, timeline: &dyn Timeline, startpoint:
         offset = 0;
     }
 
-    info!("reached end of WAL at {}", last_lsn);
-    let checkpoint_bytes = checkpoint.encode();
-    timeline.put_page_image(RelishTag::Checkpoint, 0, last_lsn, checkpoint_bytes)?;
-
-    timeline.advance_last_record_lsn(last_lsn.align());
-    timeline.checkpoint()?;
+    if last_lsn != startpoint {
+        info!(
+            "reached end of WAL at {}, updating checkpoint info",
+            last_lsn
+        );
+        let checkpoint_bytes = checkpoint.encode();
+        timeline.put_page_image(RelishTag::Checkpoint, 0, last_lsn, checkpoint_bytes)?;
+
+        timeline.advance_last_record_lsn(last_lsn);
+        timeline.checkpoint()?;
+    } else {
+        info!("no WAL to import at {}", last_lsn);
```

Comment:
We haven't advanced lsn in `import_timeline_from_postgres_datadir()`, so I think here we should call
```
timeline.advance_last_record_lsn(startpoint); // startpoint == last_lsn here
timeline.checkpoint()?;
```

to flush data on disk.

Reply:
Agreed. I thought we did that in the caller, but we don't. And it would be harder for the caller to do it, because it doesn't know how much WAL was imported. I'll put it back.

Reply:
I had to change this again, thanks to commit 6a8785379a. That added an assertion that you don't call `get_page_at_lsn()` with an LSN >= latest_record_lsn. We were doing that in two places in these import functions:
1. When importing pg_control file. We first imported it as the ControlFile relish, then immediately read it back with get_page_at_lsn() to extract the checkpoint record from it and write it as a separate relish. That was easy to fix, I changed it so that it reads the file once, and creates both relishes from that without calling get_page_at_lsn()
2. import_timeline_wal() begins by reading the old checkpoint record with get_page_at_lsn() call. That fails the assertion if you haven't advanced last record LSN before calling import_timeline_wal().

To fix 2., I put the advance_last_record_lsn() call back into import_timeline_from_postgres_datadir().